### ISSUE TITLE
Himani Recruit Completion Submission

### ISSUE THREAD

**Issue by himani-maq** [microsoft/agent-academy#927]
### Curriculum Completion Validation

- [x] I verify that I have completed the Agent Academy - Recruit curriculum.

### Solution Verification

- [x] I verify that I have uploaded a screenshot of the Agent configuration with my environment name shown into the field below.
- [x] I verify that I have uploaded a screenshot of the final solution file with my environment name shown into the field below.

### Final Solution Screenshots

<img width="1914" height="921" alt="Image" src="https://github.com/user-attachments/assets/0b2beb76-e3f6-481d-b627-f885b81d5b01" />
<img width="1912" height="916" alt="Image" src="https://github.com/user-attachments/assets/f827c8a7-dc0e-47f0-b66b-96122be4227e" />

### Key Learnings and Takeaways

Introduction to agents 
Conversational AI refers to any system that can understand, process and respond to human behaviour.
Eg chatbots on websites and virtual assistants
These mostly rely on LLMs

Large Language models:
Neural networks trained on huge amount of text 
They learn through statistical patterns of language to do the work 
	1. Training data: ingests huge terabytes of text 
	2. Tokenization: Text is broken down into smaller units called tokens, model predicts one token at a time
	3. Context window: Each LLM has a limit on how many tokens it can see at once
	4. Prompting: you interact with LLM by sending a prompt 
Retrieval augmented generation (RAG):
LLMs rely on static training data, RAG help models to look up for fresh information before composing an answer
How a RAG works :
	1. User query : user asks a questions 
	2. Retriever step: System queries knowledge sources for relevant information
	3. Augmentation: Retrieved data gets appended or prepended before sending to LLM
	4. Generation: Based in user question and knowledge sources, it then generates a response that is up to date 
Conversational vs Autonomous agents:
Conversational:
 requires two way dialog to work
Persist context across multiple turns of conversation
Can hook into external APIs or tools 
Examples:
• An agent in Microsoft Teams that answers HR policy questions.
• An agent on a public website that answer questions about your products.

Autonomous:
Can take actions on behalf of user
Uses LLM reasoning loops( plan-act-observe-replan)
Can also hook into external tools or APIs
Operate without constant human prompts, once triggered can handle multi step processes
Examples:
• An agent that generates a travel itinerary, books flights, and emails confirmations as soon as you put in a travel request in your backend system.
• A “Meeting Summarizer” agent that joins a Teams call, transcribes it in real time, and writes an executive summary to OneNote.

Agents in copilot studio:
These are some points or features we can say by which copilot helps us build agents 
Visual agent designer
Model selection
Knowledge
Tools 
Multi modal support 
Publishing and distribution

RAG:real time knowledge integration

Copilot studio fundamentals 

What are agents in copilot studio?
Agents are specialized AI assistant designed to handle specific tasks or queries
Unlike a general purpose chatbot , your agent knows company specific data, carries out real world tasks , maintains conversational contexts.

When and why to use copilot studio?
When we need domain specific knowledge 
When you want to automate multi step workflows
You need a contextual, In-tool experience

Four Building blocks of agent
Knowledge 
Tools 
Topics 
Instructions
	1. Knowledge : data and context agent uses to answer questions accurately 
	Custom instructions and context 
	Knowledge sources(Grounding data): connect agent to multiple data sources
	2. Tools (Actions): tools define what the agent can do beyond chatting
	Each action is a task the agent executes programmatically
	
	How agents work:
	
	Define Inputs and Outputs
	Combine actions into workflows
	Connect to external system 
	3. Topics
	Defines the conversational triggers or entry points for your agent. Each topic corresponds to a piece of functionality 
	Conversational Triggers[¶](https://microsoft.github.io/agent-academy/recruit/02-copilot-studio-fundamentals/#conversational-triggers)
	• A topic might be "Submit IT Ticket," "Check Vacation Balance," or "Create Sales Report."
	• Under the hood, Copilot Studio uses generative orchestration: rather than relying on exact keywords, the AI interprets user intent and picks the right topic based on a short description you provide.

        Topic descriptions
        Mapping topics to Actions

	4. Instructions:
	Prompts or style messages
	
	
	
Create a declarative agent :
Declarative agent: a specialized operative embedded directly into Microsoft 365 copilot and teams 
Business to Employee : Interactions and services that a business provides directly 
to its employees In context to agent, It refers to using the advanced capabilities of copilot studio to support and enhance the work experience of employees

Creating a declarative agent :
Using a copilot studio
Created a custom IT help desk agent using copilot studio, added a specialized prompt and published it to Microsoft copilot studio and teams to use

Creating a Solution:
Solution is like a container that groups related components of an app ( think of it as a digital briefcase that contains the agent and the artifacts)
A solution may contain 
• App
• Tables
• Flows
• Agents
• Prompts and tools 
• Connectors

Solutions are important for application lifecycle management, that helps manage the apps and agents through every step 
Traditionally solutions were made in power apps maker portal, now there is a solution explorer we do not need to switch to the traditional method 
Now we can 
Create a solution 
Set your preferred solution 
Add or remove components
Export and Import Solutions 
Create and manage solution pipelines
Git Integration 

There are two types of solutions :
Unmanaged: used during development, can easily customized as needed
Managed: used when ready to deploy, they are locked to prevent accidental changes

Think of Solutions as a toolbox. When you need to fix or build something (an agent) in a different location (environment), you gather all the necessary tools (components) and put them in your toolbox (Solution). You can then carry this toolbox to the new location (environment) and use the tools (components) to complete your work, or add new tools (components) to customize your agent or project you're building.

Why solutions :
Organized development 
Safe deployment 
Version control 
Dependency management 
Team collaboration 

Solution publisher :
Like a label which keeps track of who created or owns a solution 
When creating a solution , we must select a publisher which contains :
A prefix that gets added to all custom components (tables, flows)
Name and contact info of the person who owns the solution 

Why?
Easy identification, avoids conflicts, Supports ALM (when moving the solution through different stages Dev-> test -> Prod) , publisher keeps track of ownership and maintain consistency

Power Platform lifecycle :
	1. Creating a solution in development environment 
	2. Add components
	3. Export as a managed solution 
	4. Import to test environment 
	5. Import to production environment
	6. Apply patches, updates or upgrades

Lab: Creating a new solution :
While creating the solution ,selecting the more options will show us an option configuration page (Developers set up an HTML web resource to help users interact with their app, agent or tool where it will appear as a web page in the information section with instructions or buttons.
This page can include: 
	• Instructions for users.
	• Buttons or links for actions (e.g., “Start Setup”, “Open Help”).
	• Interactive UI elements to guide users.

Using prebuilt agents:
Prebuilt agents are agents created by Microsoft
Lab: Creating the pre-built agent 
We can just create a pre-built agent by selecting agent template 

Create agent from conversation:
What is a custom agent?
It is a chatbot or virtual assistant the you create and design in copilot studio 
To create a custom agent, we can give prompts to copilot 
Prompts are the message or instructions given to copilot about what to do 

Grounding it with knowledge:
Knowledge sources are places where your agent can find information to give better answers
When you add these resources, agent can pull in the enterprise data

What is generative orchestration ??
Means the agent uses AI to answer the user questions by combining the built in language skills with the information we provided as a source of knowledge 

Why knowledge sources matter?
Smarter answers , Less manual work, use trusted information , works with generative AI, Flexible and expandable. 

Lab : creating a custom agent from conversation 

Add new topic with trigger
a Topic is a conversation flow or scenario that defines how your agent responds to a specific user intent or question.
Types of topics:
System topics: built in and handle common events
Custom topics: handle specific tasks or questions 

What does a topic contain??
Trigger phrases: words or sentences user may use to start the topic
Conversational nodes: nodes which agent follows once the topic is triggered

Creating a Topic:
For creating this, we gave instructions
Added input output and initialization 

Add an adaptive card:
An adaptive card is a way to create Interactive and visually rich UI element that can be embedded in apps. It is a structured json object that defines the layout and content of a card
Json: Javascript object notation : format used to structure data 
Looks like a series of key value pairs enclosed within curly braces

Adaptive card designer: visual tool that helps build interactive message cards using drag and drop elements
Elements : Card elements , Card viewer, Card Structure, Element properties ,Card payload editor( raw json code behind the card)
We can do it using either javascript or formula 

Agent flow:
Used to automate repititive tasks 
Agent flows: automate tasks inside agent, use AI and keep everything in copilot studio 
Power automate flows: automate across web and apps, build workflows outside agents
These always follow a fixed path, they do the same thing every time given any input 

How they work:
Trigger: An event starts the flow
Actions: steps that follows next 

How to create :
Natural language flow 
Designer canvas

Expressions are small formulas or commands that help your agent flow work with data. You use them to calculate values, format text, make decisions, or pull specific information from inputs.

### Challenges Faced

Choosing JSON vs Formula mode

Challenge: Deciding whether to keep a static JSON card or switch to Formula (Power Fx) to render dynamic device lists.
Impact: JSON can’t loop over data (Global.VarDevices.value), causing expectations mismatch.
Fix:
Use JSON for static lists (hardcoded options).
Use Formula for dynamic lists via ForAll(Global.VarDevices.value, …).

Adaptive Card schema issues
Challenge: Missing required properties (e.g., type, version, and element IDs), or using wrong element types (e.g., items without a container).
Symptoms: Card fails to render or output variables don’t appear.
Fix: Validate against Adaptive Card schema (v1.5), ensure:
JSON{  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",  "type": "AdaptiveCard",  "version": "1  "version": "1.5",  "body": [ ... ],  "actions": [ ... ]Show more lines

Output variables not binding
Challenge: Card renders but deviceSelectionId / commentsId not created or are empty.
Fix: Ensure Input.ChoiceSet and Input.Text have unique id values that match the outputs you expect (e.g., deviceSelectionId, commentsId), and the card is in JSON mode (not Formula).

 Formula (Power Fx) Challenges 
Invalid global variable names
Challenge: Errors like “Name isn’t valid. 'VarDevices' isn’t…” or “The '.' operator cannot be used on Error values.”
Cause: Global.VarDevices not set by a prior Get items node, misspelling, or null response.
Fix:
Create the variable in a previous node using SharePoint Get items → set to Global.VarDevices.
Reference as Global.VarDevices.value (for connector responses).

ForAll invalid arguments
Challenge: “The function 'ForAll' has some invalid arguments.”
Cause: Global.VarDevices.value isn’t a table or record shape differs (e.g., fields named Title instead of Model).
Fix: Check actual column names; sample pattern:
Plain Textpowerfx isn’t fully supported. Syntax highlighting is based on Plain Text.ForAll(  Global.VarDevices.value,  {    title: If(IsBlank(ThisRecord.Model), "Unknown", ThisRecord.Model),    value: Text(ThisRecord.ID)  }Show more lines

Topic Wiring & Flow Challenges
No agent flow path
Challenge: You decided not to use an Agent flow; needed to complete logic inside the topic.
Impact: No email/SharePoint write-back; must rely on confirmation messages.
Fix:
Add Send a message node using variables {deviceSelectionId} and {commentsId}.
End with Topic management → End of Conversation.

Variable visibility & scope
Challenge: Not seeing system variables or custom outputs when inserting variables.
Fix: Use Insert variable → System → User.DisplayName for personalization, and Custom tab for adaptive card outputs.

Testing & UX Challenges
Environment/header not visible in screenshots
Challenge: Badges/forms require environment name visible; screenshots miss header.
Fix: Capture full browser window (don’t crop header). Show Solutions or Test pane with environment label.

Adaptive card submission behavior
Challenge: Submit button triggers but downstream message shows empty values.
Cause: User didn’t select a device; isRequired missing on ChoiceSet.
Fix: Set:
JSON"isRequired": true,"errorMessage": "Please select a device."``Show more lines

Friendly display (ID vs name)
Challenge: Confirmation message shows numeric ID (123) instead of readable device name.
Fix (JSON mode): Hardcode choices.title to device names in JSON; message reads the name if you store it in a variable or include it in the card.

connector/Permissions 
Challenge: Outlook/SharePoint connections prompting Allow or failing due to permissions.
Fix: Ensure you’re signed in to the right tenant; re-authenticate connectors; verify SharePoint list access.

### Final Thoughts

In this lab, I learned how to enhance a Copilot Studio topic by adding an interactive Adaptive Card using JSON. The card allowed users to select a device and provide additional comments, with outputs captured for use in the conversation flow. I also implemented a confirmation message and ended the conversation without using an agent flow, ensuring a smooth user experience. Key takeaways include understanding Adaptive Card structure, managing output variables, and designing user-friendly flows. This exercise reinforced the importance of testing and validating topic behavior for accuracy and usability.

### Badge validation form

- [x] I verify that I have completed the badge validation form.

**Comment by VictorDevRel** [comment]
Your solution has been verified, and your badge is on the way! Thank you so much for completing the course and taking the time to share your learnings. ⭐ Keep up the good work, Recruit! Check your email for further instructions
